site/public/js/testcase-output.js

```diff
@@ -0,0 +1,103 @@
+ * @param {string} gradeable_id - ID of the gradeable
+ * @param {string} who_id - ID of the submitter
+ * @param {string} index - index of test case
+ * @param {number} version - submission version
+ */
+// eslint-disable-next-line no-unused-vars
+function loadTestcaseOutput(div_name, gradeable_id, who_id, index, version = '') {
+    const orig_div_name = div_name;
+    div_name = `#${div_name}`;
+
+    const loadingTools = $(`#tc_${index}`).find('.loading-tools');
+
+    // Checks if output for this div is being loaded
+    if (loading_testcases_xml_http_requests.has(orig_div_name)) {
+        // Checks if xhr is defined and has not succeeded yet
+        // If so, abort loading
+        const xhr = loading_testcases_xml_http_requests.get(orig_div_name);
+        if (xhr && xhr.readyState !== 4) {
+            xhr.abort();
+            loading_testcases_xml_http_requests.delete(orig_div_name);
+        }
+    }
+    // Checks if test case output is visible
+    // Output is visible - Collapse test case output
+    if ($(div_name).is(':visible')) {
+        CollapseTestcaseOutput(orig_div_name, index, loadingTools);
+    }
+    // Output is not visible - Expand test case output
+    else {
+        $(`#show_char_${index}`).toggle();
+        // eslint-disable-next-line no-undef
+        const url = `${buildCourseUrl(['gradeable', gradeable_id, 'grading', 'student_output'])}?who_id=${who_id}&index=${index}&version=${version}`;
+
+        loadingTools.find('.loading-tools-show').hide();
+        loadingTools.find('.loading-tools-in-progress').show();
+
+        loading_testcases_xml_http_requests.set(orig_div_name, $.getJSON({
+            url: url,
+            async: true,
+            success: function(response) {
+                if (response.status !== 'success') {
+                    alert(`Error getting file diff: ${response.message}`);
+                    return;
+                }
+                loading_testcases_xml_http_requests.delete(orig_div_name);
+
+                $(div_name).empty();
+                $(div_name).html(response.data);
+                // eslint-disable-next-line no-undef
+                toggleDiv(orig_div_name);
+
+                loadingTools.find('.loading-tools-in-progress').hide();
+                loadingTools.find('.loading-tools-hide').show();
+                // eslint-disable-next-line no-undef
+                enableKeyToClick();
+            },
+            error: function(e) {
+                if (loading_testcases_xml_http_requests.has(orig_div_name) && loading_testcases_xml_http_requests.get(orig_div_name).readyState === 0) {
+                    loading_testcases_xml_http_requests.delete(orig_div_name);
+                    CollapseTestcaseOutput(orig_div_name, index, loadingTools);
+                    console.log('JSON Load Aborted');
+                }
+                else {
+                    alert('Could not load diff, please refresh the page and try again.');
```

Comment:
If there is a case where this `else` path gets executed wouldn't you want to call the `ColllapseTestcaseOutput` function and also delete the `request` from the map if it exists. I thought that the deleting request and call to `CollapseTestcaseoutput` should be done outside `if-else` statement.

Reply:
If this else is reached, then the test case could not be loaded due to an error and the page would need to be refreshed. Since the page will be refreshed, we would not need to remove the request from the map since when the page is changed or refreshed, global objects do not persist, so it's only relevant to the current page.

Reply:
Ah yes that makes sense